q2_moshpit/plugin_setup.py

```diff
@@ -221,19 +227,24 @@
 )
 
 T_kraken_collate_reports_in, T_kraken_collate_reports_out = TypeMap({
-    SampleData[Kraken2Reports % Properties('reads', 'contigs')]: (
+    SampleData[Kraken2Reports % Properties('reads', 'contigs', 'mags')]:
+        SampleData[Kraken2Reports % Properties('reads', 'contigs', 'mags')],
+    SampleData[Kraken2Reports % Properties('reads', 'contigs')]:
         SampleData[Kraken2Reports % Properties('reads', 'contigs')],
-    ),
-    SampleData[Kraken2Reports % Properties('reads')]: (
+    SampleData[Kraken2Reports % Properties('reads', 'mags')]:
+        SampleData[Kraken2Reports % Properties('reads', 'mags')],
+    SampleData[Kraken2Reports % Properties('contigs', 'mags')]:
+        SampleData[Kraken2Reports % Properties('contigs', 'mags')],
+    SampleData[Kraken2Reports % Properties('reads')]:
         SampleData[Kraken2Reports % Properties('reads')],
-    ),
-    SampleData[Kraken2Reports % Properties('contigs')]: (
+    SampleData[Kraken2Reports % Properties('contigs')]:
         SampleData[Kraken2Reports % Properties('contigs')],
-    )
+    SampleData[Kraken2Reports % Properties('mags')]:
+        SampleData[Kraken2Reports % Properties('mags')],
 })
```

Comment:
Hey @ebolyen, does this look correct? This was the only way to make the TypeMap work after the addition of the "mags" property. I am not exactly sure I understand why I'm otherwise getting the error telling me the options are ambiguous...

Reply:
Just fyi this should also fail **from the command line** if I'm right about qiime2/qiime2#784

Reply:
I think the reason why this might not have failed in the past was that the collate actions were called from within the `classify-kraken2` action, and never by themselves (e.g. `qiime moshpit collate-kraken2-reports`)

Reply:
Hey @misialq, yes this is correct. The reason this happens and looks like a powerset sans-null, is that the properties are a kind of intersection type. It is simultaneously reads, contigs, and mags. Without the more specific branches at the top (which short-circuit) QIIME 2 reasons that a Kraken2Reports of `reads`, `contigs`, and `mags` might exist, but it would apply equally to a branch `reads`, `contigs` OR `reads`, `mags`. So it mentions there's ambiguity in the branches and hopes you fix it exactly like you did.

Reply:
It would be nice to have a short-hand for this situation someday, because it does look pretty silly I will admit.

Reply:
Actually, you could use `TypeMatch([...])` here instead and list only the `Properties(...)` (you'll still need the silly powerset), and then you can just use the type-variable as the property for the input and output. That'll be a bit more manageable.

Reply:
It occurs to me that the `TypeMatch` of only properties (instead of `TypeMap` of full expressions) would also sidestep the issue @Sann5 found in the framework.